### ISSUE TITLE
Enhancement: evolve Tikhon into a token-efficient LLM reasoning language grounded in classical AI and modern research

### ISSUE THREAD

**Issue by AlexKay28** [AlexKay28/tahoe#17]
## Goal

Develop Tikhon into a **formal, logical, compact, concrete, human-readable reasoning language whose driving engine is an LLM**. Its primary purpose is to reduce total reasoning expenditure without a quality drop, improve stability, and evolve through measured use.

The language should express goals, knowledge, inference, uncertainty, decisions, and revision without requiring every operation to be a conventional executable algorithm. The runtime provides state, evidence, validation, and measurement around model-interpreted operations.

Related: #16 (implementation roadmap). This proposal supplies a research/design foundation and evaluation criteria for that roadmap. The examples below are proposed notation, not assertions about currently supported syntax.

## Design thesis

**Optimize useful reasoning per token, not visual brevity.** A compact trace can omit necessary computation or hide ambiguity. A successful language preserves decision-relevant distinctions and expands difficult steps on demand.

Use the accumulated experience of logic, cybernetics, cognitive architectures, knowledge representation, and modern LLM research. Historical inspiration is not evidence of LLM performance: every borrowed mechanism must become a falsifiable experiment.

### 1. Notation as a cognitive tool

**Kenneth E. Iverson — “Notation as a Tool of Thought” (1980, US).**
Source: https://www.jsoftware.com/papers/tot.htm

- Borrow composability, economy, suggestiveness, and the ability to subordinate detail.
- Introduce constructs for recurring reasoning operations rather than arbitrary abbreviations of prose.
- Provide a compact form with recoverable definitions and an expanded explanatory form.
- Test natural-language keywords against symbolic notation under each model's actual tokenizer; compare correctness, interpretation errors, prompt overhead, and total tokens.

**Leibniz — characteristica universalis / calculus ratiocinator (17th-century Europe).**

- Borrow the separation between representation of concepts and operations on those representations.
- Keep claim/relationship syntax distinct from inference/control syntax.
- Historical motivation, not a claim that a universal symbolic calculus can capture all open-ended reasoning.

### 2. Situation-driven reasoning and changing representations

**D. A. Pospelov — “Ситуационное управление: теория и практика” (1986, USSR); related work on semiotic models/systems.**

- Borrow situation recognition and selection of applicable knowledge/actions when no complete fixed model is available.
- Represent `situation → applicable knowledge → action → revised situation`.
- Study semiotic systems as inspiration for revisable representations and rules; specify our own explicit, testable evolution mechanism.
- Test changing requirements, incomplete observations, and recovery after an initial plan becomes invalid.

**Valentin Turchin — Refal, supercompilation, and “The Phenomenon of Science” (1977; USSR → US).**

- Borrow symbolic expression transformation, specialization, and higher-level control over recurring lower-level processes.
- Investigate specializing general reasoning protocols to recurring task families.
- Preserve applicability conditions when replacing a sequence with a named abstraction.
- Test whether specialization saves total cost on held-out tasks without overgeneralization. Supercompilation guarantees for programs do not automatically transfer to LLM reasoning.

**Lev Vygotsky — “Мышление и речь” / “Thinking and Speech” (1934, USSR).**

- Borrow the hypothesis that compact expression depends on shared context and recoverable references.
- Keep explicit state and emit concise updates rather than repeatedly narrating established facts.
- Test reference recovery after context truncation, checkpoint reload, and worker handoff.
- Treat inner speech as inspiration, not a literal account of transformer cognition.

### 3. Declarative knowledge and control of inference

**John McCarthy — “Programs with Common Sense” (1959, US), Advice Taker.**

- Borrow declarative goals, constraints, and knowledge that permit an intelligent engine to select actions.
- Allow partially specified problems rather than requiring a fully enumerated execution path.
- Test generalization to tasks where useful actions cannot be known before gathering evidence.

**Robert Kowalski — “Algorithm = Logic + Control” (1979, Europe).**

- Separate meaning from strategy: claims/constraints describe the problem; search order, budget, and stopping policy control reasoning.
- Compare multiple control policies over identical problem representations.
- Check that switching strategies does not silently alter goals or assumptions.

**Charles Sanders Peirce — abduction; Jon Doyle — “A Truth Maintenance System” (1979); Johan de Kleer — “An Assumption-Based TMS” (1986), US.**

- Distinguish observation, assumption, hypothesis, derivation, and refutation.
- Track support and dependency links. On retraction, invalidate or mark dependent conclusions for revalidation.
- Treat abductive explanations as hypotheses to discriminate, not deductions already proved.
- Test contradictory evidence, false initial hypotheses, and cascading revisions. Measure stale-conclusion reuse and recovery cost.

### 4. Resource-bounded reasoning

**Stuart Russell and Eric Wefald — “Do the Right Thing” (1991).**

- Treat further computation as a decision with cost and expected benefit.
- Support explicit expand, test, branch, compress, and stop decisions.
- Use cheap discriminating tests before expensive exploration when appropriate.
- Compare adaptive allocation with fixed budgets, particularly on the hardest task stratum. Avoid spending more tokens deciding whether to think than the policy saves.

### 5. Evolution through reusable abstractions

**Allen Newell — Soar and “Unified Theories of Cognition” (1990, US).**

- Borrow problem spaces, impasses, subgoals, and chunking: `impasse → subproblem → resolution → reusable rule`.
- Store applicability conditions and failure cases alongside each learned pattern.
- Compare library-enabled and library-disabled runs on unseen tasks.

**Ellis et al. — DreamCoder (2021).**
Source: https://arxiv.org/abs/2006.08381

- Borrow the loop of solving tasks, discovering useful abstractions, and improving future search with a learned library.
- Mine successful traces for candidates; also mine failures for counterexamples and overly broad conditions.
- Evaluate candidates on separate development data and promote only measured improvements; keep a final held-out test set untouched by selection.
- DreamCoder concerns program synthesis; generalizing it to open-ended LLM protocols is a research hypothesis.

**Jorma Rissanen — Minimum Description Length (1978, Europe).**

- Charge abstractions for their definition, examples, retrieval, ambiguity, and maintenance costs.
- Prefer a small useful vocabulary over uncontrolled macro accumulation.
- Measure amortized savings over a declared reuse horizon, including library/prompt overhead and repair cost.
- The proposed engineering objective is inspired by MDL, not an exact statement of its formal objective.

### 6. Direct LLM evidence and implementation mechanisms

**Xu et al. — “Chain of Draft: Thinking Faster by Writing Less” (2025).**
Source: https://arxiv.org/abs/2502.18600

- Use concise natural-language intermediate reasoning as a mandatory baseline.
- The paper reports matching/exceeding CoT accuracy with as little as 7.6% of its tokens in evaluated settings; this is not a universal saving guarantee.
- Require Tikhon to demonstrate value beyond merely asking for short reasoning.

**Pfau, Merrill, Bowman — “Let's Think Dot by Dot: Hidden Computation in Transformer Language Models” (2024).**
Source: https://arxiv.org/abs/2404.15758

- Additional tokens can supply computation independent of readable intermediate content in specially trained settings.
- Distinguish eliminating redundant narration from removing computational capacity.
- Include adaptive expansion and difficulty-stratified evaluation; do not infer internal reasoning from visible traces alone.

**Yao et al. — ReAct (2022/2023).**
Source: https://arxiv.org/abs/2210.03629

- Borrow compact interleaving of reasoning, action, observation, and update.
- Give external observations provenance and keep them distinguishable from predictions.
- Measure tool calls, redundant searches, and recovery as well as text tokens.

**Geng et al. — “Grammar-Constrained Decoding for Structured NLP Tasks without Finetuning” (2023).**
Source: https://arxiv.org/abs/2305.13971

- Investigate grammar-constrained generation for syntax and structural stability.
- Separate grammar checks, state/dependency checks, and evidence-based semantic checks.
- Compare constrained/unconstrained generation for validity, task accuracy, latency, and repair overhead. Valid syntax is not valid reasoning.

### A. Small, versioned semantic core

Candidate concepts: `goal`, `constraint`, `observe`, `assume`, `hypothesize`, `derive`, `depends`, `test`, `revise`, `retract`, `done`.

Define each construct's meaning, allowed state transition, evidence requirements where applicable, and failure behavior. Explicitly distinguish mechanically checkable semantics from model judgments.

### B. Adaptive reasoning control

Strategies select the next useful operation and control expansion, branching, evidence collection, compression, and stopping. Stable operator meanings coexist with revisable plans and beliefs.

### C. Learned protocol library

Versioned, inspectable abstractions such as `diagnose`, `compare`, `find-counterexample`, or `verify-change`, each with a definition, applicability conditions, evidence, and known failure modes. Keep the semantic core stable while testing library evolution first.

### D. Measurement and execution support

Use runtime state/evidence/replay facilities to measure and check the reasoning protocol. A sealed program or well-formed trace alone does not establish reasoning quality. Record which model, language version, library version, and control policy produced each result.

Illustrative design sketch, not current syntax:

```text
goal: restore service
constraint: preserve data
o1: observed failure began after deployment [source: incident log]
h1: hypothesis configuration mismatch [motivated-by: o1]
next: test h1 using configuration diff
on refute h1: retract h1; revalidate dependents; select alternative
done: service checks pass & data integrity verified
```

`motivated-by` intentionally does not mean logical entailment. This kind of distinction must survive compression.

### Baselines and controls

Compare the same models, task instances, tool access, and sampling settings under:

1. Ordinary natural-language instructions.
2. Concise reasoning / Chain-of-Draft-style instructions.
3. Structured natural-language state plus updates.
4. Tikhon core without learned protocols.
5. Tikhon with adaptive control.
6. Tikhon with the evolved library.

Run ablations for notation, explicit epistemic types, dependency tracking, constrained decoding, adaptive expansion, and learned abstractions. Evaluate prompt-only adoption separately from any fine-tuning experiments; charge training costs separately.

### Tasks

Use independently checkable reasoning and coding tasks, tool-use investigations, contradictory/new evidence, changed requirements, and state restoration/handoffs. Include multiple difficulty strata and model families. Use repeated trials to measure stability, not only one best run.

### Metrics

- Correctness and task completion against external checks, not self-reported success.
- Quality difference with confidence intervals and a predeclared non-inferiority margin; set that margin to zero if claiming literally no quality drop. An inconclusive comparison is not proof of equivalence.
- Total input/output tokens across all calls, including language definitions, examples, retrieval, retries, repairs, and judges.
- Provider-reported reasoning tokens where available, with accounting that avoids double-counting tokens already included in output usage. Report unavailable quantities as unobservable, not zero.
- Latency, monetary cost, tool calls, and amortized library-learning overhead.
- Syntax failures, semantic/protocol violations, stale conclusions, unnecessary steps, and variance across runs.
- Recovery cost after refuted assumptions or interrupted execution.

Predeclare the target task distribution and acceptable token/cost saving before confirmatory testing. Report quality–cost Pareto curves and difficult-task regressions, not only aggregate averages. Novel notation does not demonstrate improved reasoning by itself.

## Work packages / acceptance criteria

- [ ] Produce an annotated source/design matrix with verified bibliographic details and explicit hypothesis versus established-result labels.
- [ ] Write a semantic-core RFC, including uncertainty, scope, dependencies, retraction, and model-judgment boundaries.
- [ ] Specify compact/expanded representations and measure tokenizer behavior on target models.
- [ ] Build a baseline evaluation harness with complete usage accounting, repeated trials, and independent correctness checks.
- [ ] Run controlled core-language experiments against concise and structured natural language; publish failures as well as gains.
- [ ] Prototype adaptive expansion/stopping and dependency-aware revision; measure each contribution separately.
- [ ] Prototype protocol mining, applicability checking, versioning, and rollback using separate learning/selection/test data.
- [ ] Run transfer and robustness evaluations across task families, difficulty, and models.
- [ ] Promote mechanisms only when evidence supports their quality/cost benefit; map implementation follow-ups back to #16 and existing issues.

## Open design decisions

- Which distinctions deserve core syntax, and which belong in the protocol library?
- Which symbols are both token-efficient and reliably understood by target models?
- How much semantics can be checked without another expensive model call?
- What state must survive compression and context loss?
- When does learning a macro pay for its definition and retrieval costs?
- Which mechanisms transfer across models, and which require model-specific adaptation?

**Success means an empirically better way for LLMs to solve problems: fewer total resources, preserved quality, more stable revision and execution, and a vocabulary that improves through validated experience.**

**Comment by AlexKay28** [comment]
Delivered in commit 4ecc2bc (Wave 8). Sealed run: demo/runs/issue-17-reasoning-language (seal 4cb5afc1...63d5). docs/design/01-reasoning-language-foundation.md: design thesis, 16-row source->mechanism->Tikhon-construct table (all existence claims cited to src/ paths), 15 falsifiable experiments (E1-E15) with hypothesis/variables/measurement/baseline/threshold, per-model tokenizer notation policy, non-goals. Experiment EXECUTION and measured evidence land under #26 (benchmarks); this issue covers the design foundation and evaluation criteria per its scope.
